File: ios/Runner/KalmanFilter1D.swift
import Foundation

/// Basic 1-D Kalman filter, used here to smooth the median RSSI. Seeds
/// itself with the first measurement so it doesn't have to climb up from
/// zero. processNoise sets how fast it follows changes: higher reacts
/// quicker, lower smooths more.
final class KalmanFilter1D {

  private let measurementNoise: Double
  private let processNoise: Double
  private var estimateError: Double
  private var estimate = 0.0
  private var seeded = false

  init(measurementNoise: Double, initialEstimateError: Double, processNoise: Double) {
    precondition(measurementNoise > 0, "measurementNoise must be positive")
    precondition(initialEstimateError > 0, "initialEstimateError must be positive")
    precondition(processNoise >= 0, "processNoise must be non-negative")
    self.measurementNoise = measurementNoise
    self.estimateError = initialEstimateError
    self.processNoise = processNoise
  }

  func update(_ measurement: Double) -> Double {
    if !seeded {
      estimate = measurement
      seeded = true
      return estimate
    }
    let gain = estimateError / (estimateError + measurementNoise)
    let next = estimate + gain * (measurement - estimate)
    estimateError = (1.0 - gain) * estimateError + abs(estimate - next) * processNoise
    estimate = next
    return estimate
  }
}

File: android/app/src/main/kotlin/com/example/ble_proximity_bridge/KalmanFilter1D.kt
package com.example.ble_proximity_bridge

import kotlin.math.abs

/**
 * Basic 1-D Kalman filter, used here to smooth the median RSSI. It seeds
 * itself with the first measurement so it does not start from zero.
 * processNoise controls how quickly it follows changes: higher reacts
 * faster, lower smooths more.
 */
class KalmanFilter1D(
    private val measurementNoise: Double,
    initialEstimateError: Double,
    private val processNoise: Double,
) {
    init {
        require(measurementNoise > 0) { "measurementNoise must be positive" }
        require(initialEstimateError > 0) { "initialEstimateError must be positive" }
        require(processNoise >= 0) { "processNoise must be non-negative" }
    }

    private var estimateError = initialEstimateError
    private var estimate = 0.0
    private var seeded = false

    fun update(measurement: Double): Double {
        if (!seeded) {
            estimate = measurement
            seeded = true
            return estimate
        }
        val gain = estimateError / (estimateError + measurementNoise)
        val next = estimate + gain * (measurement - estimate)
        estimateError =
            (1.0 - gain) * estimateError + abs(estimate - next) * processNoise
        estimate = next
        return estimate
    }
}
